src/main/java/net/spy/memcached/MemcachedConnection.java

```diff
@@ -556,10 +556,10 @@ private void switchoverMemcachedReplGroup(MemcachedNode node) {
     if (group.getMasterNode() != null && group.getMasterCandidate() != null) {
       if (((ArcusReplNodeAddress) node.getSocketAddress()).isMaster()) {
         MemcachedNode masterCandidateNode = group.getMasterCandidate();
+        ((ArcusReplKetamaNodeLocator) locator).switchoverReplGroup(group);
```

Comment:
이 이슈가 예전부터 존재했던 이슈인가요?
아니면, 삼중화 지원부터 존재했던 이슈인가요?

Reply:
@jhpark816 
삼중화 이전 코드는 switchover후 slave가 한번 더 REPL_SLAVE 응답을 받을 수 있을 가능성이 있습니다. 그러나 move operation을 수행할 수 있도록 되어있었는데, 

삼중화 지원 이후로 slave가 한번 더 REPL_SLAVE 응답을 받게 될 때 candidate가 null임에 따라 move operation이 되지 않습니다.

Reply:
이왕 고치는 김에 아래와 같이 고치면 어떤가 싶네요.

```java
    if (group.getMasterNode() != null && group.getMasterCandidate() != null) {
      if (((ArcusReplNodeAddress) node.getSocketAddress()).isMaster()) {
        ((ArcusReplKetamaNodeLocator) locator).switchoverReplGroup(group);
      }
      if (node.moveOperations(group.getMasterNode()) > 0) {
         addedQueue.offer(group.getMasterNode());
      }
      queueReconnect(node, ReconnDelay.IMMEDIATE,
          "Discarded all pending reading state operation to move operations.");
    } else {
      getLogger().warn("Delay switchover because invalid group state : " + group);
    }
```

이렇게 수정하고 보니, 아래 사항이 보이네요.
moveOperations() 수행 후에, addedQueue.offer() 수행할 필요가 있나요?
- 예전엔 inputQ로 move하였기에 addedQueue에 destination node 정보를 등록하였는 데,
- 현재는 writeQ로 move하므로 addedQueue에 등록할 필요가 없는 것 같습니다.
- 현재 addedQueue에 destination node 정보를 등록하더라도 동작에 문제는 없죠?
   - addedQueue 등록된 node에 대해 처리하는 로직을 살펴보기 바랍니다. 


그리고, 추가 질문이 있습니다.
- 여기서 move 시킨 연산은 항상 1개 이상이죠?
   - 만약, move 시킨 연산이 없다면, queueReconnect() 조차도 할 필요가 없을 것 같아서요.
   - 즉, queueReconnect() 함수도 moveOperations() 결과가 0 초과인 경우에만 수행.
- SWITCHOVER or REPL_SLAVE인 경우는 move할 연산이 1개 이상이지만, ZK 이벤트로 switchover하는 경우는 move할 연산이 없을 수도 있습니다.
   - 이 경우에는 queueReconnect() 수행할 필요가 없을 것 같습니다.
- switchover 작업에서 role change하였지만 연산을 move하지 않은 상태일 수 있습니다.
  이 경우, operations 수행 순서가 아래의 이유로 보장되는 거죠?
  - work thread는 새로운 master node에게 연산을 주입하면 inputQ에만 들어가게 됨.
    즉, switchover에 의한 연산이 move되기 전에 새로운 연산이 먼저 들어올 수 있음.
  - ZK 이벤트 or 서버 응답에 의해 switchover 수행할 경우 이 작업은 IO thread가 수행하는 것이며, writeQ로 연산을 move 시키면 이 연산이 먼저 처리가 되고, inputQ에 들어온 연산은 그 이후에 처리됨.
  - 이러한 이유로 예전에 inputQ 대신에 writeQ로 move하도록 수정한 것인가요?

Reply:
@jhpark816 

>현재는 writeQ로 move하므로 addedQueue에 등록할 필요가 없는 것 같습니다.

addedQueue에서 꺼낸 노드는 fixupOps를 호출합니다. 여기서 writeQ의 op가 존재하면, 채널의 작업 유형(OP_WRITE) 등록할 수 있어, 쓰기 작업을 하기 위해서는 반드시 addedQueue의 등록 작업을 수행해주어야 합니다. 

```java
private void handleInputQueue() {
    if (!addedQueue.isEmpty()) {
      getLogger().debug("Handling queue");
      // If there's stuff in the added queue.  Try to process it.
      Collection<MemcachedNode> toAdd = new HashSet<MemcachedNode>();
      // Transfer the queue into a hashset.  There are very likely more
      // additions than there are nodes.
      Collection<MemcachedNode> todo = new HashSet<MemcachedNode>();

      MemcachedNode node;
      while ((node = addedQueue.poll()) != null) {
        todo.add(node);
      }

      // Now process the queue.
      for (MemcachedNode qa : todo) {
       ...
        qa.fixupOps(); 
      }
      addedQueue.addAll(toAdd);
    }
  }
```

```java
  public final void fixupOps() {
    // As the selection key can be changed at any point due to node
    // failure, we'll grab the current volatile value and configure it.
    SelectionKey s = sk;
    if (s != null && s.isValid()) {
      int iops = getSelectionOps();
      getLogger().debug("Setting interested opts to %d", iops);
      s.interestOps(iops);
    } else {
      getLogger().debug("Selection key is not valid.");
    }
  }

public final int getSelectionOps() {
    int rv = 0;
    if (getChannel().isConnected()) {
      if (hasReadOp()) {
        rv |= SelectionKey.OP_READ;
      }
      if (toWrite > 0 || hasWriteOp()) {
        rv |= SelectionKey.OP_WRITE;
      }
    } else {
      rv = SelectionKey.OP_CONNECT;
    }
    return rv;
  }
```

Reply:
@jhpark816 

> 여기서 move 시킨 연산은 항상 1개 이상이죠?
만약, move 시킨 연산이 없다면, queueReconnect() 조차도 할 필요가 없을 것 같아서요.
즉, queueReconnect() 함수도 moveOperations() 결과가 0 초과인 경우에만 수행.

네 여기서는 반드시 1개 이상입니다. 왜냐하면 REPL_SLAVE 응답을 받는 operation을 moving을 해주어야 하기 때문입니다. (handleRead에서 OP가 Moving 상태라면 readQ에 제거되지 않음). 그러고보니 moveOperation의 결과에 관한 if문은 여기서 굳이 필요하지 않은 것 같습니다.

> SWITCHOVER or REPL_SLAVE인 경우는 move할 연산이 1개 이상이지만, ZK 이벤트로 switchover하는 경우는 move할 연산이 없을 수도 있습니다.
이 경우에는 queueReconnect() 수행할 필요가 없을 것 같습니다.

네, updateConnection 쪽은 task로 한꺼번에 처리되고 있어서 수정하려면 고민이 필요할 듯 합니다.

Reply:
@jhpark816 

> switchover 작업에서 role change하였지만 연산을 move하지 않은 상태일 수 있습니다.
이 경우, operations 수행 순서가 아래의 이유로 보장되는 거죠?
work thread는 새로운 master node에게 연산을 주입하면 inputQ에만 들어가게 됨.
즉, switchover에 의한 연산이 move되기 전에 새로운 연산이 먼저 들어올 수 있음.
ZK 이벤트 or 서버 응답에 의해 switchover 수행할 경우 이 작업은 IO thread가 수행하는 것이며, writeQ로 연산을 move 시키면 이 연산이 먼저 처리가 되고, inputQ에 들어온 연산은 그 이후에 처리됨.
이러한 이유로 예전에 inputQ 대신에 writeQ로 move하도록 수정한 것인가요?

네 switchover시 이동되는 operation과 중간에 새로 들어오는 operation의 순서를 지키기 위해 writeQ로 move하도록 수정한 것으로 보입니다.